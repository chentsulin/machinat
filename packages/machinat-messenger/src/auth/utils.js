// @flow
/* eslint-disable import/prefer-default-export */
import type { AuthorizerRefineResult } from '@machinat/auth/types';
import type { ExtensionContext } from '../types';
import MessengerChannel from '../channel';
import MessengerUser from '../user';

export const refineExtensionContext = (
  ctx: ExtensionContext
): null | AuthorizerRefineResult => {
  if (!ctx || !ctx.page_id || !ctx.psid || !ctx.tid) {
    return null;
  }

  return {
    user: new MessengerUser(ctx.page_id, ctx.psid),
    sourceChannel: MessengerChannel.fromExtensionContext(ctx),
  };
};
