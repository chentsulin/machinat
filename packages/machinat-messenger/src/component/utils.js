import invariant from 'invariant';
import { MESSENGER_NAITVE_TYPE } from '../symbol';

export const annotateNative = component =>
  Object.assign(component, { $$native: MESSENGER_NAITVE_TYPE });

export const annotateNativeRoot = (component, entry) =>
  Object.assign(component, {
    $$entry: entry,
    $$native: MESSENGER_NAITVE_TYPE,
    $$root: true,
  });

const textReducer = (text, cur) => {
  invariant(
    typeof cur.rendered === 'string',
    `expect element at ${cur.path} can be rendered as string, got ${
      cur.element
    }`
  );
  return text + cur.rendered;
};

export const renderTextContent = (node, render, propName) => {
  const rendered = render(node, `.${propName}`);
  if (rendered === undefined) {
    return undefined;
  }

  return rendered.reduce(textReducer, '');
};

export const getRendered = result => result.rendered;
export const getElement = result => result.element;

const onlyType = (Type, result) => {
  for (let i = 0; i < result.length; i += 1) {
    if (Type !== result[i].element.type) return false;
  }
  return true;
};
export const renderOnlyType = (Type, node, render, propName) => {
  const result = render(node, `.${propName}`);
  if (result === undefined) {
    return undefined;
  }

  invariant(
    onlyType(Type, result),
    `${propName} prop should contain only ${
      Type.name
    } elements, got ${result.map(getElement)}`
  );
  return result;
};

const onlyInTypes = (types, result) => {
  for (let i = 0; i < result.length; i += 1) {
    if (types.indexOf(result[i].element.type) === -1) return false;
  }
  return true;
};
export const renderOnlyInTypes = (types, node, render, propName) => {
  const result = render(node, `.${propName}`);
  if (result === undefined) {
    return undefined;
  }

  invariant(
    onlyInTypes(types, result),
    `${propName} prop should contain only ${types
      .map(Type => Type.name)
      .join(', ')} elements, got ${result.map(getElement)}`
  );
  return result;
};
