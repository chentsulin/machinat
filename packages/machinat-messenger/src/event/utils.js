export const mixin = (...descriptors) =>
  Object.defineProperties({}, Object.assign({}, ...descriptors));

export const makeEvent = (name, proto) => raw =>
  Object.create(proto, {
    raw: {
      enumerable: true,
      value: raw,
    },
    type: {
      enumerable: true,
      value: name,
    },
  });
