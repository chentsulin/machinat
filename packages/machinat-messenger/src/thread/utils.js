// @flow
import type { ActionWithoutPause } from 'machinat-base/types';
import type { MessengerActionValue, MessengerComponent } from '../types';

export const isMessage = (
  element: $ElementType<
    ActionWithoutPause<MessengerActionValue, MessengerComponent>,
    'element'
  >
): boolean %checks =>
  typeof element !== 'object' ||
  typeof element.type === 'string' ||
  element.type.$$entry === undefined;

export const appendField = (body: string, key: string, value: string) =>
  // eslint-disable-next-line prefer-template
  (body === '' ? body : body + '&') + key + '=' + encodeURIComponent(value);

export const appendFields = (prefix: string, fields: { [string]: any }) => {
  let body = prefix;

  for (const key of Object.keys(fields)) {
    const fieldValue = fields[key];

    if (fieldValue !== undefined) {
      body = appendField(
        body,
        key,
        typeof fieldValue === 'string' ? fieldValue : JSON.stringify(fieldValue)
      );
    }
  }

  return body;
};
