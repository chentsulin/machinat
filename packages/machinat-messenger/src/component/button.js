import valuesOfAssertedTypes from '@machinat/core/utils/valuesOfAssertedTypes';

import { asPartComponent } from '../utils';
import { GenericTemplate } from './template';

const URLButton = async ({
  title,
  url,
  heightRatio,
  extensions,
  fallbackURL,
  hideShareButton,
}) => ({
  type: 'web_url',
  title,
  url,
  webview_height_ratio: heightRatio,
  messenger_extensions: extensions,
  fallback_url: fallbackURL,
  webview_share_button: hideShareButton ? 'hide' : undefined,
});
const __URLButton = asPartComponent(URLButton);

const PostbackButton = async ({ title, payload }) => ({
  type: 'postback',
  title,
  payload,
});
const __PostbackButton = asPartComponent(PostbackButton);

const getGenericTemplateValues = valuesOfAssertedTypes(() => [GenericTemplate]);

const ShareButton = async ({ children }, render) => {
  const sharedSegments = await render(children, '.children');

  let sharedContent;
  if (sharedSegments !== null) {
    sharedContent = getGenericTemplateValues(sharedSegments)[0].message;
  }

  return {
    type: 'element_share',
    share_contents: sharedContent,
  };
};
const __ShareButton = asPartComponent(ShareButton);

const BuyButton = async ({
  title,
  payload,
  currency,
  paymentType,
  isTest,
  merchantName,
  requestedInfo,
  priceList,
}) => ({
  type: 'payment',
  title,
  payload,
  payment_summary: {
    currency,
    payment_type: paymentType,
    is_test_payment: isTest,
    merchant_name: merchantName,
    requested_user_info: requestedInfo,
    price_list: priceList,
  },
});
const __BuyButton = asPartComponent(BuyButton);

const CallButton = async ({ title, number }) => ({
  type: 'phone_number',
  title,
  number,
});
const __CallButton = asPartComponent(CallButton);

const LoginButton = async ({ url }) => ({
  type: 'account_link',
  url,
});
const __LoginButton = asPartComponent(LoginButton);

const ACCOUNT_UNLINK_TYPE = { type: 'account_unlink' };
const LogoutButton = async () => ACCOUNT_UNLINK_TYPE;
const __LogoutButton = asPartComponent(LogoutButton);

const GamePlayButton = async ({ title, payload, playerId, contextId }) => ({
  type: 'game_play',
  title,
  payload,
  game_metadata: {
    player_id: playerId,
    context_id: contextId,
  },
});
const __GamePlayButton = asPartComponent(GamePlayButton);

export {
  __URLButton as URLButton,
  __PostbackButton as PostbackButton,
  __ShareButton as ShareButton,
  __BuyButton as BuyButton,
  __CallButton as CallButton,
  __LoginButton as LoginButton,
  __LogoutButton as LogoutButton,
  __GamePlayButton as GamePlayButton,
};
