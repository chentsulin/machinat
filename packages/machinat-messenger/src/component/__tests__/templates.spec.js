import moxy from 'moxy';
import Machinat from '@machinat/core';
import Renderer from '@machinat/core/renderer';
import { isNativeElement } from '@machinat/core/utils/isXxx';

import {
  GenericItem,
  GenericTemplate,
  ListTemplate,
  ButtonTemplate,
  MediaTemplate,
  OpenGraphTemplate,
  ReceiptTemplate,
  ReceiptItem,
} from '../template';
import { URLButton, PostbackButton, CallButton } from '../button';

const generalComponentDelegator = moxy(() => null);
const renderer = new Renderer('messenger', generalComponentDelegator);

test.each([
  GenericTemplate,
  ListTemplate,
  ButtonTemplate,
  MediaTemplate,
  OpenGraphTemplate,
  ReceiptTemplate,
])('attribute of %p', Template => {
  expect(typeof Template).toBe('function');
  expect(isNativeElement(<Template />)).toBe(true);
  expect(Template.$$platform).toBe('messenger');
});

test.each([GenericItem, ReceiptItem])('attribute of %p', Item => {
  expect(typeof Item).toBe('function');
  expect(isNativeElement(<Item />)).toBe(true);
  expect(Item.$$platform).toBe('messenger');
});

describe('GenericTemplate', () => {
  const items = [
    <GenericItem
      title="foo"
      imageURL="http://foo.bar/image"
      buttons={[<URLButton title="check" url="http://xxx.yy.z" />]}
    />,
    <GenericItem
      title="foo"
      subtitle="bar"
      imageURL="http://foo.bar/image"
      defaultAction={<URLButton title="TITLE!" url="http://foo.bar/" />}
      buttons={[
        <URLButton title="check" url="http://xxx.yy.z" />,
        <PostbackButton title="more" payload="_MORE_" />,
        <CallButton title="call us" number="+12345678" />,
      ]}
    />,
  ];

  it('match snapshot', async () => {
    await expect(
      renderer.render(
        <>
          <GenericTemplate>{items}</GenericTemplate>
          <GenericTemplate imageAspectRatio="square" sharable>
            {items}
          </GenericTemplate>
        </>
      )
    ).resolves.toMatchSnapshot();
  });
});

describe('ListTemplate', () => {
  it('match snapshot', async () => {
    const items = [
      <GenericItem
        title="foo"
        imageURL="http://foo.bar/image"
        buttons={[<URLButton title="check" url="http://xxx.yy.z" />]}
      />,
      <GenericItem
        title="foo"
        subtitle="bar"
        imageURL="http://foo.bar/image"
        defaultAction={<URLButton title="TITLE!" url="http://foo.bar/" />}
        buttons={[<PostbackButton title="more" payload="_MORE_" />]}
      />,
    ];

    await expect(
      renderer.render(
        <>
          <ListTemplate>{items}</ListTemplate>
          <ListTemplate
            button={<PostbackButton title="more" payload="_MORE_" />}
            imageAspectRatio="square"
            sharable
          >
            {items}
          </ListTemplate>
        </>
      )
    ).resolves.toMatchSnapshot();
  });
});

describe('ButtonTemplate', () => {
  it('match snapshot', async () => {
    generalComponentDelegator.mock.fake((node, path) => [
      { type: 'text', value: node.props.children, node, path },
    ]);

    const buttons = [
      <URLButton title="check" url="http://xxx.yy.z" />,
      <PostbackButton title="more" payload="_MORE_" />,
      <CallButton title="call us" number="+12345678" />,
    ];

    await expect(
      renderer.render(
        <>
          <ButtonTemplate buttons={buttons}>
            hello button template
          </ButtonTemplate>
          <ButtonTemplate buttons={buttons} sharable>
            <i>foo</i>
            <b>bar</b>
            <del>baz</del>
          </ButtonTemplate>
        </>
      )
    ).resolves.toMatchSnapshot();
  });

  it('render empty string as text if children is empty', async () => {
    const template = (
      <ButtonTemplate buttons={<PostbackButton title="foo" payload="foo" />} />
    );
    await expect(renderer.render(template)).resolves.toEqual([
      {
        type: 'unit',
        value: {
          message: {
            attachment: {
              type: 'template',
              payload: {
                text: '',
                template_type: 'button',
                buttons: [{ payload: 'foo', title: 'foo', type: 'postback' }],
              },
            },
          },
        },
        node: template,
        path: '$',
      },
    ]);
  });

  it('throw if <br /> contained in text prop', async () => {
    generalComponentDelegator.mock.fake((node, path) => [
      { type: 'break', node, path },
    ]);

    expect(
      renderer.render(
        <ButtonTemplate>
          foo
          <br />
          bar
        </ButtonTemplate>
      )
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"non-textual node <br /> received, only textual nodes allowed"`
    );
  });

  it('throw if non-text node contained in text prop', async () => {
    generalComponentDelegator.mock.fake((node, path) => [
      { type: 'part', value: { some: 'stranger' }, node, path },
    ]);

    expect(
      renderer.render(
        <ButtonTemplate>
          foo
          <stranger />
          bar
        </ButtonTemplate>
      )
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"non-textual node <stranger /> received, only textual nodes allowed"`
    );
  });
});

describe('MediaTemplate', () => {
  it('match snapshot', async () => {
    const buttons = [
      <URLButton title="check" url="http://xxx.yy.z" />,
      <PostbackButton title="more" payload="_MORE_" />,
      <CallButton title="call us" number="+12345678" />,
    ];

    await expect(
      renderer.render(
        <>
          <MediaTemplate type="image" url="http://..." buttons={buttons} />
          <MediaTemplate
            type="video"
            attachmentId="__ID__"
            sharable
            buttons={buttons}
          />
        </>
      )
    ).resolves.toMatchSnapshot();
  });
});

describe('OpenGraphTemplate', () => {
  it('match snapshot', async () => {
    const buttons = [
      <URLButton title="check" url="http://xxx.yy.z" />,
      <PostbackButton title="more" payload="_MORE_" />,
      <CallButton title="call us" number="+12345678" />,
    ];
    expect(
      renderer.render(
        <>
          <OpenGraphTemplate url="http://..." buttons={buttons} />
          <OpenGraphTemplate url="http://..." sharablebuttons={buttons} />
        </>
      )
    ).resolves.toMatchSnapshot();
  });
});

describe('ReceiptTemplate', () => {
  it('match snapshot', async () => {
    await expect(
      renderer.render(
        <>
          <ReceiptTemplate
            recipientName="John Doe"
            orderNumber="12345"
            currency="USD"
            paymentMethod="Visa 2345"
            orderURL="http://what.a.shop"
            summary={{ total_cost: 56.14 }}
          />

          <ReceiptTemplate
            recipientName="John Doe"
            orderNumber="12345"
            currency="USD"
            paymentMethod="Visa 2345"
            orderURL="http://what.a.shop"
            summary={{ total_cost: 56.14 }}
          >
            <ReceiptItem title="Buzz Lightyear" price={9.99} />
            <ReceiptItem title="Woody" price={9.99} />
            <ReceiptItem title="Slinky Dog" price={9.99} />
          </ReceiptTemplate>

          <ReceiptTemplate
            sharable
            recipientName="John Doe"
            merchantName="AI"
            orderNumber="12345"
            currency="USD"
            paymentMethod="Visa 2345"
            orderURL="http://what.a.shop"
            timestamp="1428444852"
            address={{
              street_1: '1 Hacker Way',
              street_2: 'Somewhere...',
              city: 'Menlo Park',
              postal_code: '94025',
              state: 'CA',
              country: 'US',
            }}
            summary={{
              subtotal: 75.0,
              shipping_cost: 4.95,
              total_tax: 6.19,
              total_cost: 56.14,
            }}
            adjustments={[
              { name: 'New Customer Discount', amount: 20 },
              { name: '$10 Off Coupon', amount: 10 },
            ]}
          >
            <ReceiptItem
              title="A robot!"
              subtitle="It's real!"
              quantity={999}
              price={99.99}
              currency="USD"
              imageURL="http://i.robot/avatar"
            />
          </ReceiptTemplate>
        </>
      )
    ).resolves.toMatchSnapshot();
  });

  it('accept Date object for timestamp prop', async () => {
    expect(
      (await renderer.render(
        <ReceiptTemplate timestamp={new Date(1535622297000)} />
      )).value
    ).toEqual(
      (await renderer.render(<ReceiptTemplate timestamp="1535622297" />)).value
    );
  });
});
