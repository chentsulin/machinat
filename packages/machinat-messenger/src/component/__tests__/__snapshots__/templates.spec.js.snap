// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonTemplate match snapshot 1`] = `
Array [
  Array [
    Object {
      "node": <ButtonTemplate
        sharable={true}
        text="abc"
      >
        <Machinat.Fragment>
          <URLButton
            title="check"
            url="http://xxx.yy.z"
          />
          <PostbackButton
            payload="_MORE_"
            title="more"
          />
          <CallButton
            number="+12345678"
            title="call us"
          />
        </Machinat.Fragment>
      </ButtonTemplate>,
      "path": "$",
      "type": "unit",
      "value": Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "buttons": Array [
                "__RENDERED_BUTTON_OBJ_1__",
                "__RENDERED_BUTTON_OBJ_2__",
                "__RENDERED_BUTTON_OBJ_3__",
              ],
              "sharable": true,
              "template_type": "button",
              "text": "
__RENDERED_TEXT_1__
__RENDERED_TEXT_2__
__RENDERED_TEXT_3__",
            },
            "type": "template",
          },
        },
      },
    },
  ],
  Array [
    Object {
      "node": <ButtonTemplate
        text={
          Array [
            <i>
              foo
            </i>,
            <b>
              bar
            </b>,
            <del>
              baz
            </del>,
          ]
        }
      >
        <Machinat.Fragment>
          <URLButton
            title="check"
            url="http://xxx.yy.z"
          />
          <PostbackButton
            payload="_MORE_"
            title="more"
          />
          <CallButton
            number="+12345678"
            title="call us"
          />
        </Machinat.Fragment>
      </ButtonTemplate>,
      "path": "$",
      "type": "unit",
      "value": Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "buttons": Array [
                "__RENDERED_BUTTON_OBJ_1__",
                "__RENDERED_BUTTON_OBJ_2__",
                "__RENDERED_BUTTON_OBJ_3__",
              ],
              "sharable": undefined,
              "template_type": "button",
              "text": "
__RENDERED_TEXT_1__
__RENDERED_TEXT_2__
__RENDERED_TEXT_3__",
            },
            "type": "template",
          },
        },
      },
    },
  ],
]
`;

exports[`GenericItem match snapshot 1`] = `
Array [
  Array [
    Object {
      "node": <GenericItem
        title="foo"
      />,
      "path": "$",
      "type": "part",
      "value": Object {
        "buttons": undefined,
        "default_action": undefined,
        "image_url": undefined,
        "subtitle": undefined,
        "title": "foo",
      },
    },
  ],
  Array [
    Object {
      "node": <GenericItem
        subtitle="bar"
        title="foo"
      />,
      "path": "$",
      "type": "part",
      "value": Object {
        "buttons": undefined,
        "default_action": undefined,
        "image_url": undefined,
        "subtitle": "bar",
        "title": "foo",
      },
    },
  ],
  Array [
    Object {
      "node": <GenericItem
        imageURL="http://foo.bar/image"
        title="foo"
      />,
      "path": "$",
      "type": "part",
      "value": Object {
        "buttons": undefined,
        "default_action": undefined,
        "image_url": "http://foo.bar/image",
        "subtitle": undefined,
        "title": "foo",
      },
    },
  ],
  Array [
    Object {
      "node": <GenericItem
        title="foo"
      >
        <Machinat.Fragment>
          <URLButton
            title="check"
            url="http://xxx.yy.z"
          />
          <PostbackButton
            payload="_MORE_"
            title="more"
          />
          <CallButton
            number="+12345678"
            title="call us"
          />
        </Machinat.Fragment>
      </GenericItem>,
      "path": "$",
      "type": "part",
      "value": Object {
        "buttons": Array [
          "__RENDERED_BUTTON_OBJ_1__",
          "__RENDERED_BUTTON_OBJ_2__",
          "__RENDERED_BUTTON_OBJ_3__",
        ],
        "default_action": undefined,
        "image_url": undefined,
        "subtitle": undefined,
        "title": "foo",
      },
    },
  ],
  Array [
    Object {
      "node": <GenericItem
        defaultAction={
          <URLButton
            extensions={true}
            fallbackURL="http://foo.baz/login"
            heightRatio="compact"
            hideShareButton={true}
            title="TITLE!"
            url="http://foo.bar/"
          />
        }
        title="foo"
      />,
      "path": "$",
      "type": "part",
      "value": Object {
        "buttons": undefined,
        "default_action": Object {
          "fallback_url": "http://foo.baz/login",
          "messenger_extensions": true,
          "type": "web_url",
          "url": "http://foo.bar/",
          "webview_height_ratio": "compact",
          "webview_share_button": "hide",
        },
        "image_url": undefined,
        "subtitle": undefined,
        "title": "foo",
      },
    },
  ],
  Array [
    Object {
      "node": <GenericItem
        defaultAction={
          Object {
            "fallback_url": "http://foo.baz/login",
            "messenger_extensions": true,
            "type": "web_url",
            "url": "http://foo.bar/",
            "webview_height_ratio": "compact",
            "webview_share_button": "hide",
          }
        }
        title="foo"
      />,
      "path": "$",
      "type": "part",
      "value": Object {
        "buttons": undefined,
        "default_action": Object {
          "fallback_url": "http://foo.baz/login",
          "messenger_extensions": true,
          "type": "web_url",
          "url": "http://foo.bar/",
          "webview_height_ratio": "compact",
          "webview_share_button": "hide",
        },
        "image_url": undefined,
        "subtitle": undefined,
        "title": "foo",
      },
    },
  ],
  Array [
    Object {
      "node": <GenericItem
        defaultAction={
          Object {
            "fallback_url": "http://foo.baz/login",
            "messenger_extensions": true,
            "type": "web_url",
            "url": "http://foo.bar/",
            "webview_height_ratio": "compact",
            "webview_share_button": "hide",
          }
        }
        imageURL="http://foo.bar/image"
        subtitle="bar"
        title="foo"
      >
        <Machinat.Fragment>
          <URLButton
            title="check"
            url="http://xxx.yy.z"
          />
          <PostbackButton
            payload="_MORE_"
            title="more"
          />
          <CallButton
            number="+12345678"
            title="call us"
          />
        </Machinat.Fragment>
      </GenericItem>,
      "path": "$",
      "type": "part",
      "value": Object {
        "buttons": Array [
          "__RENDERED_BUTTON_OBJ_1__",
          "__RENDERED_BUTTON_OBJ_2__",
          "__RENDERED_BUTTON_OBJ_3__",
        ],
        "default_action": Object {
          "fallback_url": "http://foo.baz/login",
          "messenger_extensions": true,
          "type": "web_url",
          "url": "http://foo.bar/",
          "webview_height_ratio": "compact",
          "webview_share_button": "hide",
        },
        "image_url": "http://foo.bar/image",
        "subtitle": "bar",
        "title": "foo",
      },
    },
  ],
]
`;

exports[`GenericTemplate match snapshot 1`] = `
Array [
  Array [
    Object {
      "node": <GenericTemplate>
        <Machinat.Fragment>
          <GenericItem
            subtitle="bar"
            title="foo1"
          />
          <GenericItem
            subtitle="baz"
            title="foo2"
          />
          <GenericItem
            imageURL="http://foo.bar/image"
            title="foo3"
          />
        </Machinat.Fragment>
      </GenericTemplate>,
      "path": "$",
      "type": "unit",
      "value": Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "elements": Array [
                "__RENDERED_GENERIC_ITEM_OBJ_1__",
                "__RENDERED_GENERIC_ITEM_OBJ_2__",
                "__RENDERED_GENERIC_ITEM_OBJ_3__",
              ],
              "image_aspect_ratio": undefined,
              "sharable": undefined,
              "template_type": "generic",
            },
            "type": "template",
          },
        },
      },
    },
  ],
  Array [
    Object {
      "node": <GenericTemplate
        imageAspectRatio="square"
        sharable={true}
      >
        <Machinat.Fragment>
          <GenericItem
            subtitle="bar"
            title="foo1"
          />
          <GenericItem
            subtitle="baz"
            title="foo2"
          />
          <GenericItem
            imageURL="http://foo.bar/image"
            title="foo3"
          />
        </Machinat.Fragment>
      </GenericTemplate>,
      "path": "$",
      "type": "unit",
      "value": Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "elements": Array [
                "__RENDERED_GENERIC_ITEM_OBJ_1__",
                "__RENDERED_GENERIC_ITEM_OBJ_2__",
                "__RENDERED_GENERIC_ITEM_OBJ_3__",
              ],
              "image_aspect_ratio": "square",
              "sharable": true,
              "template_type": "generic",
            },
            "type": "template",
          },
        },
      },
    },
  ],
]
`;

exports[`ListTemplate match snapshot 1`] = `
Array [
  Array [
    Object {
      "node": <ListTemplate>
        <Machinat.Fragment>
          <GenericItem
            subtitle="bar"
            title="foo1"
          />
          <GenericItem
            subtitle="baz"
            title="foo2"
          />
          <GenericItem
            imageURL="http://foo.bar/image"
            title="foo3"
          />
        </Machinat.Fragment>
      </ListTemplate>,
      "path": "$",
      "type": "unit",
      "value": Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "buttons": undefined,
              "elements": Array [
                "__RENDERED_GENERIC_ITEM_OBJ_1__",
                "__RENDERED_GENERIC_ITEM_OBJ_2__",
                "__RENDERED_GENERIC_ITEM_OBJ_3__",
              ],
              "sharable": undefined,
              "template_type": "list",
              "top_element_style": undefined,
            },
            "type": "template",
          },
        },
      },
    },
  ],
  Array [
    Object {
      "node": <ListTemplate
        button={
          <URLButton
            title="check"
            url="http://xxx.yy.z"
          />
        }
      >
        <Machinat.Fragment>
          <GenericItem
            subtitle="bar"
            title="foo1"
          />
          <GenericItem
            subtitle="baz"
            title="foo2"
          />
          <GenericItem
            imageURL="http://foo.bar/image"
            title="foo3"
          />
        </Machinat.Fragment>
      </ListTemplate>,
      "path": "$",
      "type": "unit",
      "value": Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "buttons": Array [
                "__RENDERED_BUTTON_OBJ_1__",
              ],
              "elements": Array [
                "__RENDERED_GENERIC_ITEM_OBJ_1__",
                "__RENDERED_GENERIC_ITEM_OBJ_2__",
                "__RENDERED_GENERIC_ITEM_OBJ_3__",
              ],
              "sharable": undefined,
              "template_type": "list",
              "top_element_style": undefined,
            },
            "type": "template",
          },
        },
      },
    },
  ],
  Array [
    Object {
      "node": <ListTemplate
        button={
          <URLButton
            title="check"
            url="http://xxx.yy.z"
          />
        }
        imageAspectRatio="square"
        sharable={true}
      >
        <Machinat.Fragment>
          <GenericItem
            subtitle="bar"
            title="foo1"
          />
          <GenericItem
            subtitle="baz"
            title="foo2"
          />
          <GenericItem
            imageURL="http://foo.bar/image"
            title="foo3"
          />
        </Machinat.Fragment>
      </ListTemplate>,
      "path": "$",
      "type": "unit",
      "value": Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "buttons": Array [
                "__RENDERED_BUTTON_OBJ_1__",
              ],
              "elements": Array [
                "__RENDERED_GENERIC_ITEM_OBJ_1__",
                "__RENDERED_GENERIC_ITEM_OBJ_2__",
                "__RENDERED_GENERIC_ITEM_OBJ_3__",
              ],
              "sharable": true,
              "template_type": "list",
              "top_element_style": undefined,
            },
            "type": "template",
          },
        },
      },
    },
  ],
]
`;

exports[`MediaTemplate match snapshot 1`] = `
Array [
  Array [
    Object {
      "node": <MediaTemplate
        type="image"
        url="http://..."
      >
        <Machinat.Fragment>
          <URLButton
            title="check"
            url="http://xxx.yy.z"
          />
          <PostbackButton
            payload="_MORE_"
            title="more"
          />
          <CallButton
            number="+12345678"
            title="call us"
          />
        </Machinat.Fragment>
      </MediaTemplate>,
      "path": "$",
      "type": "unit",
      "value": Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "elements": Array [
                Object {
                  "attachment_id": undefined,
                  "buttons": Array [
                    "__RENDERED_BUTTON_OBJ_1__",
                    "__RENDERED_BUTTON_OBJ_2__",
                    "__RENDERED_BUTTON_OBJ_3__",
                  ],
                  "media_type": "image",
                  "url": "http://...",
                },
              ],
              "sharable": undefined,
              "template_type": "media",
            },
            "type": "template",
          },
        },
      },
    },
  ],
  Array [
    Object {
      "node": <MediaTemplate
        attachmentId="__ID__"
        sharable={true}
        type="video"
      >
        <Machinat.Fragment>
          <URLButton
            title="check"
            url="http://xxx.yy.z"
          />
          <PostbackButton
            payload="_MORE_"
            title="more"
          />
          <CallButton
            number="+12345678"
            title="call us"
          />
        </Machinat.Fragment>
      </MediaTemplate>,
      "path": "$",
      "type": "unit",
      "value": Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "elements": Array [
                Object {
                  "attachment_id": "__ID__",
                  "buttons": Array [
                    "__RENDERED_BUTTON_OBJ_1__",
                    "__RENDERED_BUTTON_OBJ_2__",
                    "__RENDERED_BUTTON_OBJ_3__",
                  ],
                  "media_type": "video",
                  "url": undefined,
                },
              ],
              "sharable": true,
              "template_type": "media",
            },
            "type": "template",
          },
        },
      },
    },
  ],
]
`;

exports[`OpenGraphTemplate match snapshot 1`] = `
Array [
  Array [
    Object {
      "node": <OpenGraphTemplate
        url="http://..."
      >
        <Machinat.Fragment>
          <URLButton
            title="check"
            url="http://xxx.yy.z"
          />
          <PostbackButton
            payload="_MORE_"
            title="more"
          />
          <CallButton
            number="+12345678"
            title="call us"
          />
        </Machinat.Fragment>
      </OpenGraphTemplate>,
      "path": "$",
      "type": "unit",
      "value": Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "elements": Array [
                Object {
                  "buttons": Array [
                    "__RENDERED_BUTTON_OBJ_1__",
                    "__RENDERED_BUTTON_OBJ_2__",
                    "__RENDERED_BUTTON_OBJ_3__",
                  ],
                  "url": "http://...",
                },
              ],
              "sharable": undefined,
              "template_type": "open_graph",
            },
            "type": "template",
          },
        },
      },
    },
  ],
  Array [
    Object {
      "node": <OpenGraphTemplate
        sharable={true}
        url="http://..."
      >
        <Machinat.Fragment>
          <URLButton
            title="check"
            url="http://xxx.yy.z"
          />
          <PostbackButton
            payload="_MORE_"
            title="more"
          />
          <CallButton
            number="+12345678"
            title="call us"
          />
        </Machinat.Fragment>
      </OpenGraphTemplate>,
      "path": "$",
      "type": "unit",
      "value": Object {
        "message": Object {
          "attachment": Object {
            "payload": Object {
              "elements": Array [
                Object {
                  "buttons": Array [
                    "__RENDERED_BUTTON_OBJ_1__",
                    "__RENDERED_BUTTON_OBJ_2__",
                    "__RENDERED_BUTTON_OBJ_3__",
                  ],
                  "url": "http://...",
                },
              ],
              "sharable": true,
              "template_type": "open_graph",
            },
            "type": "template",
          },
        },
      },
    },
  ],
]
`;

exports[`ReceiptItem match snapshot 1`] = `
Array [
  Object {
    "node": <ReceiptItem
      currency="USD"
      imageURL="http://i.robot/avatar"
      price={99.99}
      quantity={999}
      subtitle="It's real!"
      title="A robot!"
    />,
    "path": "$",
    "type": "part",
    "value": Object {
      "currency": "USD",
      "image_url": "http://i.robot/avatar",
      "price": 99.99,
      "quantity": 999,
      "subtitle": "It's real!",
      "title": "A robot!",
    },
  },
]
`;

exports[`ReceiptTemplate match snapshot 1`] = `
Array [
  Object {
    "node": <ReceiptTemplate
      address={
        Object {
          "city": "Menlo Park",
          "country": "US",
          "postal_code": "94025",
          "state": "CA",
          "street_1": "1 Hacker Way",
          "street_2": "",
        }
      }
      adjustments={
        Array [
          Object {
            "amount": 20,
            "name": "New Customer Discount",
          },
          Object {
            "amount": 10,
            "name": "$10 Off Coupon",
          },
        ]
      }
      currency="USD"
      orderNumber="12345"
      orderURL="http://what.a.shop"
      paymentMethod="Visa 2345"
      recipientName="John Doe"
      sharable={true}
      summary={
        Object {
          "shipping_cost": 4.95,
          "subtotal": 75,
          "total_cost": 56.14,
          "total_tax": 6.19,
        }
      }
      timestamp="1428444852"
    >
      <ReceiptItem
        title="Buzz Lightyear"
      />
      <ReceiptItem
        title="Woody"
      />
      <ReceiptItem
        title="Slinky Dog"
      />
    </ReceiptTemplate>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "address": Object {
              "city": "Menlo Park",
              "country": "US",
              "postal_code": "94025",
              "state": "CA",
              "street_1": "1 Hacker Way",
              "street_2": "",
            },
            "adjustments": Array [
              Object {
                "amount": 20,
                "name": "New Customer Discount",
              },
              Object {
                "amount": 10,
                "name": "$10 Off Coupon",
              },
            ],
            "currency": "USD",
            "elements": Array [
              "__RENDERED_RECEIPT_TEMPLATE_ITEM_OBJ_1__",
              "__RENDERED_RECEIPT_TEMPLATE_ITEM_OBJ_2__",
              "__RENDERED_RECEIPT_TEMPLATE_ITEM_OBJ_3__",
            ],
            "order_number": "12345",
            "order_url": "http://what.a.shop",
            "payment_method": "Visa 2345",
            "recipient_name": "John Doe",
            "sharable": true,
            "summary": Object {
              "shipping_cost": 4.95,
              "subtotal": 75,
              "total_cost": 56.14,
              "total_tax": 6.19,
            },
            "template_type": "receipt",
            "timestamp": "1428444852",
          },
          "type": "template",
        },
      },
    },
  },
]
`;
